src/clib/pioc.c

```diff
@@ -468,104 +468,110 @@ int PIOc_Init_Intracomm(const MPI_Comm comp_comm, const int num_iotasks,
+    /* Ensure that settings for number of computation tasks, number
+     * of IO tasks, and the stride are reasonable. */
+    if (iosys->num_comptasks == 1 && num_iotasks * ustride > 1)
+    {
+        // This is a serial run with a bad configuration. Set up a single task.
+        fprintf(stderr, "PIO_TP PIOc_Init_Intracomm reset stride and tasks.\n");
+        iosys->num_iotasks = 1;
+        ustride = 1;
+    }
+    
+    if (iosys->num_iotasks < 1 || iosys->num_iotasks * ustride > iosys->num_comptasks)
+    {
+        fprintf(stderr, "PIO_TP PIOc_Init_Intracomm error\n");
+        fprintf(stderr, "num_iotasks=%d, ustride=%d, num_comptasks=%d\n", num_iotasks,
+                ustride, iosys->num_comptasks);
+        return PIO_EBADID;
+    }
 
-        /* For the tasks that are doing IO, get their rank. */
-        if (iosys->ioproc)
-            CheckMPIReturn(MPI_Comm_rank(iosys->io_comm, &(iosys->io_rank)),__FILE__,__LINE__);
-        else
-            iosys->io_rank = -1;
-        LOG((3, "iosys->io_comm = %d iosys->io_rank = %d", iosys->io_comm, iosys->io_rank));
+    /* Create an array that holds the ranks of the tasks to be used for IO. */
+    if (!(iosys->ioranks = calloc(sizeof(int), iosys->num_iotasks)))
+        return PIO_ENOMEM;
+    for (int i = 0; i < iosys->num_iotasks; i++)
+    {
+        iosys->ioranks[i] = (base + i * ustride) % iosys->num_comptasks;
+        if (iosys->ioranks[i] == iosys->comp_rank)
+            iosys->ioproc = true;
+    }
+    iosys->ioroot = iosys->ioranks[0];
+
+    for (int i = 0; i < iosys->num_iotasks; i++)
+        LOG((3, "iosys->ioranks[%d] = %d", i, iosys->ioranks[i]));
+
+    /* Create an MPI info object. Actually we don't want to do
+     * that. We immediately overwrite the handle with MPI_INFO_NULL,
+     * and lose the chance to ever do a free on the created info
+     * object.*/
+    /* CheckMPIReturn(MPI_Info_create(&iosys->info),__FILE__,__LINE__); */
+    iosys->info = MPI_INFO_NULL;
+
+    /* The task that has an iomaster value of MPI_ROOT will be the
+     * root of the IO communicator. */
+    if (iosys->comp_rank == iosys->ioranks[0])
+        iosys->iomaster = MPI_ROOT;
+
+    /* Create a group for the computation tasks. */
+    if ((mpierr = MPI_Comm_group(iosys->comp_comm, &iosys->compgroup)))
+        return check_mpi(NULL, mpierr, __FILE__, __LINE__);
+
+    /* Create a group for the IO tasks. */
+    if ((mpierr = MPI_Group_incl(iosys->compgroup, iosys->num_iotasks, iosys->ioranks,
+                                 &iosys->iogroup)))
+        return check_mpi(NULL, mpierr, __FILE__, __LINE__);        
+
+    /* Create an MPI communicator for the IO tasks. */
+    if ((mpierr = MPI_Comm_create(iosys->comp_comm, iosys->iogroup, &iosys->io_comm)))
+        return check_mpi(NULL, mpierr, __FILE__, __LINE__);                
+
+    /* For the tasks that are doing IO, get their rank within the IO communicator. */
+    if (iosys->ioproc)
```

Comment:
I know, it's very odd. But when I change that one, all tests break.

So I thought I would take the low-hanging fruit and get the rest of these changes merged. Then I will swing back around and investigate this one.